### PR TITLE
Generalize machines to address #39

### PR DESCRIPTION
**Pull request by alang9**
Just want to request comments on this change. The most important changes are in `Data.Machine.Type`

### CODE REVIEW

File: src/Data/Machine/Type.hs [hunk 1]
```diff
@@ -54,81 +63,129 @@ module Data.Machine.Type
...
+-- newtype MachineT m k o = MachineT { runMachineT :: m (Step k o (MachineT m k o)) }
 
 -- | A 'Machine' reads from a number of inputs and may yield results before stopping.
 --
 -- A 'Machine' can be used as a @'MachineT' m@ for any @'Monad' m@.
 type Machine k o = forall m. Monad m => MachineT m k o
 
+machineT :: (Functor m) => MStep m (Is o) k o -> MachineT m k o
+machineT m = unsafeCoerce $ \_ -> m
+{-# INLINE machineT #-}
```
Comment (treeowl) [inline src/Data/Machine/Type.hs]: Whoa... What's this with `unsafeCoerce`? That *certainly* shouldn't appear without very strong justification for

1. Why it's necessary, and
2. Why it's safe.

When answering question 1, please consider using rewrite rules instead. That is, if you can calculate a proof that `a :~: b` but don't want to do it at runtime, you can use `RULES` to replace the call to the proof with `Refl`.

Comment (alang9) [reply src/Data/Machine/Type.hs]: I thought this improved the benchmark performance, but the gain seems small. I've pushed a fix

**Comment by alang9** [comment]
I want to point out 2 things:

1. the performance seems worse in the benchmark. In particular, the scan, fold and zip benchmarks look quite bad
2. With this definition of MachineT:
```
type MachineT m k o = TranslateT m (Is o) k
```
We are not able to write instances for `Functor`, `Monad`, `Applicative`, `Foldable`. We could newtype `MachineT` to allow for these instances, but that seems more annoying.

**Comment by alang9** [comment]
Another thing I want to mention is that I chose the type argument order for `TranslateT` so that it can allow `Profunctor`-like functions

```
lmap1 :: (forall t. j t -> i t) -> TranslateT m i k -> TranslateT m j k
rmap1 :: (forall t. k t -> l t) -> TranslateT m j k -> TranslateT m j l
```

But this means that it's kind of flipped compared to the order in `MachineT`.
Also, when composing 2 `TranslateT`s, like `f . g`. `g` is the more downstream `TranslateT`, which is kind of opposite from the direction of function composition.

Any thoughts about whether the order should be flipped?

**Comment by YoEight** [comment]
Is this PR still ongoing?

**Comment by YoEight** [comment]
I'm closing. Feel free to ping if you need that MR to move forward.

**Pull request closed without merging**